src/cargo/core/resolver/encode.rs

```diff
@@ -408,68 +416,98 @@ impl EncodableResolve {
     }
 }
 
-fn build_path_deps(ws: &Workspace<'_>) -> CargoResult<HashMap<String, SourceId>> {
-    // If a crate is **not** a path source, then we're probably in a situation
-    // such as `cargo install` with a lock file from a remote dependency. In
-    // that case we don't need to fixup any path dependencies (as they're not
-    // actually path dependencies any more), so we ignore them.
-    let members = ws
-        .members()
-        .filter(|p| p.package_id().source_id().is_path())
-        .collect::<Vec<_>>();
-
-    let mut ret = HashMap::new();
-    let mut visited = HashSet::new();
-    for member in members.iter() {
-        ret.insert(
-            member.package_id().name().to_string(),
-            member.package_id().source_id(),
-        );
-        visited.insert(member.package_id().source_id());
-    }
-    for member in members.iter() {
-        build_pkg(member, ws, &mut ret, &mut visited);
+struct PathDeps {
+    path_deps: HashMap<String, Vec<PackageId>>,
+}
+
+impl PathDeps {
+    /// Get a best-effort path dependency with given constraints.
+    ///
+    /// If no exact version match, pickup a random version for patch updating trial
+    fn get_path_dep(&self, name: &str, version: &str) -> Option<&PackageId> {
+        let pkg_version = semver::Version::parse(version).ok()?;
+        self.path_deps.get(name).and_then(|versions| {
+            versions
+                .iter()
+                .find(|package_id| {
+                    package_id.version().cmp_precedence(&pkg_version) == std::cmp::Ordering::Equal
+                })
+                .or_else(|| versions.last())
```

Comment:
Use `last()` for behaviour consistency with old cargo. Old cargo use HashMap and insert for path deps construction, therefore the last inserted package was picked.